python/ray/data/dataset.py

```diff
@@ -445,13 +453,29 @@ def transform(
             for start in range(0, total_rows, max_batch_size):
                 # Build a block for each batch.
                 end = min(total_rows, start + max_batch_size)
-                # Make sure to copy if slicing to avoid the Arrow serialization
-                # bug where we include the entire base view on serialization.
-                view = block.slice(start, end, copy=batch_size is not None)
+                view = block.slice(start, end, copy=not zero_copy_batch)
                 # Convert to batch format.
                 view = BlockAccessor.for_block(view).to_batch_format(batch_format)
 
-                applied = batch_fn(view, *fn_args, **fn_kwargs)
+                try:
+                    applied = batch_fn(view, *fn_args, **fn_kwargs)
+                except ValueError as e:
+                    read_only_msgs = [
+                        "assignment destination is read-only",
+                        "buffer source array is read-only",
+                    ]
+                    err_msg = str(e)
+                    if any(msg in err_msg for msg in read_only_msgs):
+                        raise ValueError(
```

Comment:
I think we thought about that but thought it would be too much magic? The main concerns that pop up in my mind are:
1. This would assume that the UDFs are idempotent.
2. The UDFs may take a long time per batch and then fail at the end (e.g. at the end-write), which would have a performance impact.
3. This would require us to maintain state across batches.
4. Log messages may be missed, or may spam the user if there are many map tasks.

I'm leaning towards doing this, however, since I think the UX benefit outweighs those concerns. What do y'all think? cc @ericl @c21

Reply:
This is much less magical than many other things we do (like auto tune, sampling size estimation etc.), because there is just one resolution and is known by the implementation.

I think #1 is a real concern, others seem fine. At Ray level, objects are immutable from shared memory, so if the UDF is running in-memory computation, it should be fine. But if it's writing to external system, that might be an issue.

Reply:
I think that those 4 concerns are cumulatively significant enough that I don't think that we want to do this, even (2) could be disqualifying since the common case could be that e.g. GPU inference takes nearly twice as long because of this transparent retry-with-copy after performing inference and trying to store the result in the batch dataframe at the end. cc @ericl @matthewdeng @c21 for thoughts, but I think this is too spooky.